app/controllers/authenticate_controller.rb

```diff
@@ -223,6 +212,39 @@ def handle_authentication_error(err)
     end
   end
 
+  def audit_authn_error(err)
+    Audit.logger.log(
+      ::Audit::Event::Authn::Authenticate.new(
+        authenticator_name: authenticator_input.authenticator_name,
+        service: authenticator_input.webservice,
+        role_id: audit_role_id,
+        client_ip: authenticator_input.client_ip,
+        success: false,
+        error_message: err.message
+      )
+    )
+  end
+
+  def authenticator_input
```

Comment:
i think this can stay in the same place it's implemented now at the top. Anyway, let's have it next to the other authenticator_input methods